src/opt/pre_scheduling_impl.cpp

```diff
@@ -0,0 +1,334 @@
+                visited[def_node] = true;
+            }
+        }
+        if(ins->outputs().empty())
+        {
+            exit_nodes.push_back(&node);
+        }
+        node.ins         = ins;
+        node.ins_ndx     = ndx++;
+        instr2_node[ins] = &node;
+    }
+    int size = exit_nodes.size();
+    if(size > 1)
+    {
+        std::sort(exit_nodes.begin(), exit_nodes.end(), compare_exit_nodes);
+    }
+}
+
+// Do topology sort according to accumulated weight.  Identify critial paths.
+// Schedule nodes into streams.  Reorder instructions according to topological
+// order and annoate streams and events in the instructions.
+//
+void pre_scheduling_impl::reorder()
+{
+    std::list<dag_node*> sorted_nodes;
+    std::stack<dag_node*> stack;
+    std::priority_queue<dag_node*, std::vector<dag_node*>, weighted_topology_ordering> child_queue;
+    std::unordered_map<dag_node*, bool> visited;
+    std::unordered_map<dag_node*, bool> dequeued;
+
+    for(auto&& node : exit_nodes)
+    {
+        stack.push(node);
+        node->partition = partition_info.create_partition();
+        partition_info.add_weight(node);
+        while(!stack.empty())
+        {
+            auto cur = stack.top();
+            if(dequeued.find(cur) != dequeued.end())
+            {
+                stack.pop();
+                continue;
+            }
+            else if((visited.find(cur) != visited.end()) || cur->ins->inputs().empty())
+            {
+                stack.pop();
+                sorted_nodes.push_back(cur);
+                dequeued[cur] = true;
+                continue;
+            }
+            // sort child nodes.
+            for(auto&& arg : cur->ins->inputs())
+            {
+                dag_node* child_node = instr2_node[arg];
+                if(dequeued.find(child_node) == dequeued.end())
+                {
+                    child_queue.push(child_node);
+                }
+            }
+
+            // Last item in queue is on critical path.
+            while(!child_queue.empty())
+            {
+                dag_node* child = child_queue.top();
+                stack.push(child);
+                child_queue.pop();
+                if(child->weight_sum < min_partition_threshold)
+                    child->partition = cur->partition;
+                else if(!child_queue.empty())
+                    child->partition = partition_info.create_partition();
+                else
+                {
+                    cur->first_child = child;
+                    child->partition = cur->partition;
+                }
+                partition_info.add_weight(child);
+            }
+            visited[cur] = true;
+        }
+    }
+
+#ifdef MIGRAPHX_DEBUG_OPT
+    MIGRAPHX_DEBUG(dump("---After weighted topology sort---"));
+    MIGRAPHX_DEBUG(dump(sorted_nodes));
+#endif
+    schedule(sorted_nodes);
+    splice(sorted_nodes);
+    annotate(sorted_nodes);
+
+#ifdef MIGRAPHX_DEBUG_OPT
+    verify();
```

Comment:
Verify should be enabled by a boolean, similar to how it is enabled for memory_coloring. This way we can enable for the tests.